deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SessionManager.java

```diff
@@ -0,0 +1,210 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.websocket.Session;
+
+/**
+ * This class manage sessions of agents corresponding to agent ids.
+ *
+ * @since 1.0.0
+ */
+public class SessionManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(SessionManager.class);
+
+    private static SessionManager sessionManager = new SessionManager();
+    private static volatile Map<String, Session> agentSessions = new HashMap<>();
+    private static volatile Map<String, Agent> agents = new HashMap<>();
+    private static volatile Map<String, OperationResponse> operationResponses = new HashMap<>();
+
+    private SessionManager() {
+    }
+
+    /**
+     * Return instance of {@link SessionManager}.
+     *
+     * @return the instance of {@link SessionManager}.
+     */
+    public static SessionManager getInstance() {
+        return sessionManager;
+    }
+
+    /**
+     * Create agent session entry to store {@link Session} object against to the agent id.
+     *
+     * @param agentId      - Id of the agent.
+     * @param agentSession - {@link Session} belongs to the agent.
+     */
+    public synchronized void createAgentSession(String agentId, Session agentSession) {
+        Agent agent = new Agent(agentId);
+        String provider = agent.getProvider();
+        String region = agent.getRegion();
+        String instanceId = agent.getInstanceId();
+        if (provider != null && region != null && instanceId != null) {
+            agent.setInstanceName(getInstanceName(provider, region, instanceId));
+            agentSessions.put(agentId, agentSession);
+            agents.put(agentId, agent);
+        }
+    }
+
+    /**
+     * Remove {@link Session} entry when agent leaves.
+     *
+     * @param agentId - Id of the agent.
+     */
+    public synchronized void removeAgentSession(String agentId) {
+        agentSessions.remove(agentId);
+        agents.remove(agentId);
+    }
+
+    /**
+     * Get the agent by specifying the test plan id and instance name.
+     *
+     * @param testPlanId   - Test plan id which spawned the agent.
+     * @param instanceName - Name of the Instance which contains the agent.
+     * @return the unique agent if exists, null otherwise.
+     */
+    public Agent getAgent(String testPlanId, String instanceName) {
+        Optional<Map.Entry<String, Agent>> agentOptional = agents.entrySet().stream()
+                .filter(entry -> {
+                    Agent agent = entry.getValue();
+                    if (agent != null && testPlanId != null && instanceName != null) {
+                        return testPlanId.equals(agent.getTestPlanId()) && instanceName.equals(agent.getInstanceName());
+                    }
+                    return false;
+                }).findFirst();
+        return agentOptional.map(Map.Entry::getValue).orElse(null);
```

Comment:
Why not return Optional directly?

Reply:
In getAgent method, we are intended to return an Agent object. The optional contains Map.Entry<String, Agent> object.

Reply:
Yes, so how about Optional<Agent>? Optional is recommended to be used for return types.